torch/csrc/distributed/rpc/PythonRpcHandler.cpp

```diff
@@ -0,0 +1,95 @@
+#include <torch/csrc/distributed/rpc/PythonRpcHandler.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+PyObject* PythonRpcHandler::module_ = nullptr;
+PyObject* PythonRpcHandler::runUDFFunction_ = nullptr;
+PyObject* PythonRpcHandler::loadResultFunction_ = nullptr;
+bool PythonRpcHandler::pyInitializerCalled_ = false;
+
+void PythonRpcHandler::init() {
+  if (!Py_IsInitialized()) {
+    Py_Initialize();
+    pyInitializerCalled_ = true;
+  }
+  module_ = PyImport_ImportModule("torch.distributed.internal_rpc_utils");
+  if (module_ == nullptr) {
+    throw std::runtime_error("import internal_rpc_utils module is NULL");
+    PyErr_Print();
+  }
+  runUDFFunction_ = PyObject_GetAttrString(module_, "run_python_udf_internal");
+  if (runUDFFunction_ == nullptr) {
+    throw std::runtime_error("import run_python_udf_internal is NULL");
+    PyErr_Print();
+  }
+  loadResultFunction_ = PyObject_GetAttrString(
+    module_, "load_python_udf_result_internal");
+  if (loadResultFunction_ == nullptr) {
+      throw std::runtime_error(
+        "import load_python_udf_result_internal is NULL");
+      PyErr_Print();
+  }
+}
+
+void PythonRpcHandler::cleanUp() {
+  Py_DECREF(module_);
+  Py_DECREF(runUDFFunction_);
+  Py_DECREF(loadResultFunction_);
+  if(pyInitializerCalled_) {
+    Py_Finalize();
+  }
+}
+
+std::vector<char> PythonRpcHandler::generatePythonUDFResult(
+  const Message& request) {
+  PyGILState_STATE gstate;
+  gstate = PyGILState_Ensure();
+  auto pargs =
+      Py_BuildValue("(y#)", request.payload().data(), request.payload().size());
+  if (pargs == nullptr) {
+    throw std::runtime_error("python function args is NULL");
+    PyErr_Print();
+  }
+  auto pres = PyEval_CallObject(runUDFFunction_, pargs);
```

Comment:
If the UDF makes some IO calls (e.g., another sync RPC), it will block one thread on the callee side (pointed out by @xush6528 and @pietern). In the short term, we can bump up the number of recv threads on `RpcAgent`, but that does not completely solve the problem. Consider two workers A and B. A makes an RPC call to B (ping), in that RPC call B makes another RPC back to A (pong), and A-B do nested ping-pong 100 times before exiting. We don't want this to block 100 threads on both sides. To solve this problem, we might need to expose a responder/FutureResult concept to Python land. For example, the UDF can take a `dist.FutureResult()` argument. The 100 nested ping-pong can be added as callbacks to asynchronous RPC results (`FutureMessage` objects). Then the last pong on B calls `futureResult.markCompleted(value)`, which answers the very first RPC from A.

This of course does NOT need to be included in this PR, and I think itself is complex enough and would require some design discussion. Does this make sense? @zhaojuanmao @pritamdamania87 @pietern @xush6528 @satgera

Reply:
@mrshenli that makes sense, I will follow up on this design and work on it in follow up PR